File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        // 1. find the node 
        // 2. check if no child or one child exists return accordingly 
        // 3. if both child exists return inorder successor and replace
        // that with root value then from start searching from root->right for 
        // successor value when that value is found that would be delete (thats duplicate )

        return delNode(root,key);
    }

    TreeNode* delNode(TreeNode* node,int& key){
        if(!node) return NULL; 
        if(node->val>key) {
            node->left=delNode(node->left,key);
        }
        else if(node->val<key){
            node->right=delNode(node->right,key);
        }
        else{
            if(!node->left && !node->right){
                delete node;
                return NULL; 
            } 
            if(!node->left || !node->right){
                TreeNode* res = node->left?node->left:node->right;
                delete node;
                return res;
            }
            // find inorder success; 
            TreeNode* successor = node->right; 
            while(successor->left) successor = successor->left; 
            node->val = successor->val; 
            node->right = deleteNode(node->right,successor->val);

        }
        return node;
    }

};

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def searchDel(self,parent,node,key): 
        if node is None: 
            return None 
        if node.val == key: 
            rst=node.right
            lst=node.left
            prev,inorders = node.right,node.right 
            if inorders is None: 
                return lst
            while inorders.left:
                prev=inorders
                inorders=inorders.left 
            if prev is not inorders:
                prev.left = inorders.right 
            if rst is not inorders:
                inorders.right = rst 
            if lst is not inorders:
                inorders.left = lst
            return inorders 
        if node.val>key: 
            node.left=self.searchDel(node,node.left,key)
        else:
            node.right=self.searchDel(node,node.right,key)
        return node

    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        # find the node 
        # find inorder successor 
        # if IS has a right child prev -> left = right childe 
        # that will be swapp
        dummy = TreeNode(-1) 
        dummy.right=self.searchDel(dummy,root,key)
        return dummy.right